File: code/test-utils/src/main/java/com/adobe/marketing/mobile/util/TestNetworkService.java
/*
  Copyright 2021 Adobe. All rights reserved.
  This file is licensed to you under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License. You may obtain a copy
  of the License at http://www.apache.org/licenses/LICENSE-2.0
  Unless required by applicable law or agreed to in writing, software distributed under
  the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
  OF ANY KIND, either express or implied. See the License for the specific language
  governing permissions and limitations under the License.
*/

package com.adobe.marketing.mobile.util;

import com.adobe.marketing.mobile.services.HttpConnecting;
import com.adobe.marketing.mobile.services.Log;
import com.adobe.marketing.mobile.services.NetworkCallback;
import com.adobe.marketing.mobile.services.NetworkRequest;
import com.adobe.marketing.mobile.services.Networking;
import com.adobe.marketing.mobile.services.TestableNetworkRequest;
import com.adobe.marketing.mobile.util.ADBCountDownLatch;
import com.adobe.marketing.mobile.util.TestConstants;

import java.io.ByteArrayInputStream;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

public class TestNetworkService implements Networking {

	private static final String LOG_SOURCE = "TestNetworkService";
	private final Map<TestableNetworkRequest, List<TestableNetworkRequest>> receivedTestableNetworkRequests;
	private final Map<TestableNetworkRequest, HttpConnecting> responseMatchers;
	private final Map<TestableNetworkRequest, ADBCountDownLatch> expectedTestableNetworkRequests;
	private final ExecutorService executorService; // simulating the async network service
	private Integer delayedResponse = 0;

	private static final HttpConnecting defaultResponse = new HttpConnecting() {
		@Override
		public InputStream getInputStream() {
			return new ByteArrayInputStream("".getBytes());
		}

		@Override
		public InputStream getErrorStream() {
			return null;
		}

		@Override
		public int getResponseCode() {
			return 200;
		}

		@Override
		public String getResponseMessage() {
			return "";
		}

		@Override
		public String getResponsePropertyValue(String responsePropertyKey) {
			return null;
		}

		@Override
		public void close() {}
	};

	public TestNetworkService() {
		receivedTestableNetworkRequests = new HashMap<>();
		responseMatchers = new HashMap<>();
		expectedTestableNetworkRequests = new HashMap<>();
		executorService = Executors.newCachedThreadPool();
	}

	public void reset() {
		Log.trace(TestConstants.LOG_TAG, LOG_SOURCE, "Reset received and expected network requests.");
		receivedTestableNetworkRequests.clear();
		responseMatchers.clear();
		expectedTestableNetworkRequests.clear();
		delayedResponse = 0;
	}

	public void setResponseConnectionFor(
		final TestableNetworkRequest request,
		final HttpConnecting responseConnection
	) {
		responseMatchers.put(request, responseConnection);
	}

	public void setExpectedNetworkRequest(final TestableNetworkRequest request, final int count) {
		expectedTestableNetworkRequests.put(request, new ADBCountDownLatch(count));
	}

	public Map<TestableNetworkRequest, ADBCountDownLatch> getExpectedNetworkRequests() {
		return expectedTestableNetworkRequests;
	}

	public List<TestableNetworkRequest> getReceivedNetworkRequestsMatching(final TestableNetworkRequest request) {
		for (Map.Entry<TestableNetworkRequest, List<TestableNetworkRequest>> requests : receivedTestableNetworkRequests.entrySet()) {
			if (requests.getKey().equals(request)) {
				return requests.getValue();
			}
		}

		return Collections.emptyList();
	}

	public boolean isNetworkRequestExpected(final TestableNetworkRequest request) {
		return expectedTestableNetworkRequests.containsKey(request);
	}

	public boolean awaitFor(final TestableNetworkRequest request, final int timeoutMillis) throws InterruptedException {
		for (Map.Entry<TestableNetworkRequest, ADBCountDownLatch> expected : expectedTestableNetworkRequests.entrySet()) {
			if (expected.getKey().equals(request)) {
				return expected.getValue().await(timeoutMillis, TimeUnit.MILLISECONDS);
			}
		}

		return true;
	}

	public void enableDelayedResponse(final Integer delaySec) {
		if (delaySec < 0) {
			return;
		}

		delayedResponse = delaySec;
	}

	@Override
	public void connectAsync(NetworkRequest networkRequest, NetworkCallback resultCallback) {
		Log.trace(
			TestConstants.LOG_TAG,
			LOG_SOURCE,
			"Received connectUrlAsync to URL '%s' and HttpMethod '%s'.",
			networkRequest.getUrl(),
			networkRequest.getMethod().name()
		);

		executorService.submit(() -> {
			HttpConnecting response = setNetworkRequest(
				new TestableNetworkRequest(
					networkRequest.getUrl(),
					networkRequest.getMethod(),
					networkRequest.getBody(),
					networkRequest.getHeaders(),
					networkRequest.getConnectTimeout(),
					networkRequest.getReadTimeout()
				)
			);

			if (resultCallback != null) {
				if (delayedResponse > 0) {
					try {
						Thread.sleep(delayedResponse * 1000);
					} catch (InterruptedException e) {
						e.printStackTrace();
					}
				}

				resultCallback.call(response == null ? defaultResponse : response);
			}
		});
	}

	/**
	 * Add the network request to the list of received requests. Returns the matching response, or
	 * the default response if no matching response was found.
	 */
	private HttpConnecting setNetworkRequest(TestableNetworkRequest networkRequest) {
		if (!receivedTestableNetworkRequests.containsKey(networkRequest)) {
			receivedTestableNetworkRequests.put(networkRequest, new ArrayList<>());
		}

		receivedTestableNetworkRequests.get(networkRequest).add(networkRequest);

		HttpConnecting response = getMatchedResponse(networkRequest);
		countDownExpected(networkRequest);

		return response == null ? defaultResponse : response;
	}

	private void countDownExpected(final TestableNetworkRequest request) {
		for (Map.Entry<TestableNetworkRequest, ADBCountDownLatch> expected : expectedTestableNetworkRequests.entrySet()) {
			if (expected.getKey().equals(request)) {
				expected.getValue().countDown();
			}
		}
	}

	private HttpConnecting getMatchedResponse(final TestableNetworkRequest request) {
		for (Map.Entry<TestableNetworkRequest, HttpConnecting> responses : responseMatchers.entrySet()) {
			if (responses.getKey().equals(request)) {
				return responses.getValue();
			}
		}

		return null;
	}
}

File: code/test-utils/src/main/java/com/adobe/marketing/mobile/util/TestNetworkService.kt
/*
  Copyright 2021 Adobe. All rights reserved.
  This file is licensed to you under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License. You may obtain a copy
  of the License at http://www.apache.org/licenses/LICENSE-2.0
  Unless required by applicable law or agreed to in writing, software distributed under
  the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
  OF ANY KIND, either express or implied. See the License for the specific language
  governing permissions and limitations under the License.
*/

package com.adobe.marketing.mobile.util

import com.adobe.marketing.mobile.services.HttpConnecting
import com.adobe.marketing.mobile.services.Log
import com.adobe.marketing.mobile.services.TestableNetworkRequest
import java.io.ByteArrayInputStream
import java.io.InputStream
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit


class TestNetworkService {
    private val sentTestableNetworkRequests: MutableMap<TestableNetworkRequest, MutableList<TestableNetworkRequest>>
    private val networkResponses: MutableMap<TestableNetworkRequest, HttpConnecting>
    private val expectedTestableNetworkRequests: MutableMap<TestableNetworkRequest, ADBCountDownLatch>

    companion object {
        private const val LOG_SOURCE = "TestNetworkService"
    }

    init {
        sentTestableNetworkRequests = HashMap()
        networkResponses = HashMap()
        expectedTestableNetworkRequests = HashMap()
    }

    fun reset() {
        Log.trace(
            TestConstants.LOG_TAG,
            LOG_SOURCE,
            "Reset received and expected network requests."
        )
        sentTestableNetworkRequests.clear()
        networkResponses.clear()
        expectedTestableNetworkRequests.clear()
    }

    fun setResponseConnectionFor(
        request: TestableNetworkRequest,
        responseConnection: HttpConnecting
    ) {
        networkResponses[request] = responseConnection
    }

    fun setExpectedNetworkRequest(request: TestableNetworkRequest, count: Int) {
        expectedTestableNetworkRequests[request] = ADBCountDownLatch(count)
    }

    fun getExpectedNetworkRequests(): Map<TestableNetworkRequest, ADBCountDownLatch> {
        return expectedTestableNetworkRequests
    }

    fun getReceivedNetworkRequestsMatching(request: TestableNetworkRequest): List<TestableNetworkRequest> {
        for ((key, value) in sentTestableNetworkRequests) {
            if (key == request) {
                return value
            }
        }
        return emptyList()
    }

    fun isNetworkRequestExpected(request: TestableNetworkRequest): Boolean {
        return expectedTestableNetworkRequests.containsKey(request)
    }

    @Throws(InterruptedException::class)
    fun awaitFor(request: TestableNetworkRequest, timeoutMillis: Int): Boolean {
        for ((key, value) in expectedTestableNetworkRequests) {
            if (key == request) {
                return value.await(timeoutMillis.toLong(), TimeUnit.MILLISECONDS)
            }
        }
        return true
    }

    fun recordSentNetworkRequest(networkRequest: TestableNetworkRequest) {
        Log.trace(TestConstants.LOG_TAG,
            LOG_SOURCE,
            "Received connectAsync to URL ${networkRequest.url} and HTTPMethod ${networkRequest.method}")

        val equalNetworkRequest = sentTestableNetworkRequests.entries.firstOrNull { (key, _) ->
            key == networkRequest
        }

        if (equalNetworkRequest != null) {
            sentTestableNetworkRequests[equalNetworkRequest.key]?.add(networkRequest)
        } else {
            sentTestableNetworkRequests[networkRequest] = mutableListOf(networkRequest)
        }
    }

    /**
     * Records a mock network request and retrieves a matched response.
     *
     * This function adds the provided [networkRequest] to the list of sent testable network requests.
     *
     * @param networkRequest The mock network request to be recorded.
     * @param defaultResponse The default response to be returned if no matched response is found.
     * @return The matched response for the provided [networkRequest] or the [defaultResponse] if none is found.
     */
    fun setMockNetworkRequest(networkRequest: TestableNetworkRequest, defaultResponse: HttpConnecting): HttpConnecting {
        if (!sentTestableNetworkRequests.containsKey(networkRequest)) {
            sentTestableNetworkRequests[networkRequest] = ArrayList()
        }
        sentTestableNetworkRequests[networkRequest]!!.add(networkRequest)
        val response = getMatchedResponse(networkRequest)
        countDownExpected(networkRequest)
        return response ?: defaultResponse
    }

    fun countDownExpected(request: TestableNetworkRequest) {
        for ((key, value) in expectedTestableNetworkRequests) {
            if (key == request) {
                value.countDown()
            }
        }
    }

    private fun getMatchedResponse(request: TestableNetworkRequest): HttpConnecting? {
        for ((key, value) in networkResponses) {
            if (key == request) {
                return value
            }
        }
        return null
    }
}